src/core/qgsproject.cpp

```diff
@@ -2064,6 +2075,121 @@ QList<QgsMapLayer *> QgsProject::mapLayersByName( const QString &layerName ) con
   return mLayerStore->mapLayersByName( layerName );
 }
 
+bool QgsProject::unzip()
+{
+  return unzip( mArchive->filename() );
+}
+
+bool QgsProject::unzip( const QString &filename )
+{
+  clearError();
+  std::unique_ptr<QgsArchive> archive( new QgsArchive() );
+
+  // unzip the archive
+  if ( !archive->unzip( filename ) )
+  {
+    setError( tr( "Unable to unzip file '%1'" ).arg( filename ) );
+    return false;
+  }
+
+  // test if zip provides a .qgs file
+  if ( archive->projectFile().isEmpty() )
+  {
+    setError( tr( "Zip archive does not provide a project file" ) );
+    return false;
+  }
+
+  // read the project file
+  mUnzipping = true;
+  mPathResolverBaseName = filename;
+  if ( ! read( archive->projectFile() ) )
+  {
+    mPathResolverBaseName = QString();
+    mUnzipping = false;
+    setError( tr( "Cannot read unzipped qgs project file" ) );
+    return false;
+  }
+
+  // keep the archive
+  mUnzipping = false;
+  mPathResolverBaseName = QString();
+  mArchive.reset( archive.release() );
+
+  return true;
+}
+
+bool QgsProject::zip()
+{
+  if ( unzipped() )
+    return zip( mArchive->filename() );
+
+  return false;
+}
+
+bool QgsProject::zip( const QString &filename )
+{
+  clearError();
+
+  // save the current project in a temporary .qgs file
+  std::unique_ptr<QgsArchive> archive( new QgsArchive() );
+  const QString baseName = QFileInfo( filename ).baseName();
+  const QString qgsFileName = QString( "%1.qgs" ).arg( baseName );
+  QFile qgsFile( QDir( archive->dir() ).filePath( qgsFileName ) );
+
+  bool writeOk;
+  if ( qgsFile.open( QIODevice::WriteOnly ) )
```

Comment:
I think this needs ` | QIODevice::Truncate ` also (at least that was required in Qt4, not sure if it's been fixed in Qt5)